mesa_geo/geospace.py

```diff
@@ -56,11 +67,16 @@ def add_agents(self, agents):
         if isinstance(agents, GeoAgent):
             agent = agents
             if hasattr(agent, "geometry"):
+                if not self.crs.is_exact_same(agent.crs):
+                    agent.to_crs(self.crs)
```

Comment:
I assume this `agent.to_crs` mutates the `crs` of `agent`. I wonder if there should be info message printed to the user about it. Implicit conversion/side effect under the hood might cause problem, or maybe not in this case?

Reply:
Yeah it could possibly lead to serious problems depending on what the users want to do. The  `epsg:3857` crs is a reasonable default for `GeoSpace`, but may not be suitable for every project.

As @Corvince mentioned in the README file:

> The unit for the distance depends on the coordinate reference system (CRS) of the GeoSpace. Since we did not specify the CRS, Mesa-Geo defaults to the 'Web Mercator' projection (in meters). If you want to do some serious measurements you should always set an appropriate CRS, since the accuracy of Web Mercator declines with distance from the equator.  We can achieve this by initializing the AgentCreator and the GeoSpace with the `crs` keyword  `crs="epsg:2163"`. Mesa-Geo then transforms all coordinates from the GeoJSON geographic coordinates into the set crs.

I also thought about issuing a warning when doing such conversion. The thing is, it is common to bulk load `GeoAgent` from Shapefiles or GeoDataFrame. Then the user will be flooded with all these conversion warnings. What's even worse is when a user actually wants to use the default and do such conversion.

So, I simply assumed that users would know what they're doing.

Reply:
Another option is to raise an exception if `not self.crs.is_exact_same(agent.crs)`, but otherwise forces the auto-conversion if the user specifies `auto_convert_crs=True`. The error message can also tell users about the existence of the `auto_convert_crs` optional argument. Though, I don't have enough data points of the use case to be able to tell if this will simply make things inconvenient.

Reply:
My worry is that, if users have multiple Shapefiles (quite common) but each Shapefile is of different `crs` (probably not very often but could happen), then they will have to set `auto_convert_crs` to `True` in order to avoid errors, regardless they use default `crs` or not (unless of course, `crs` is manually set to the correct one when initializing `AgentCreator`).

I don't know enough use cases either. Let me check with my colleagues and advisor and get back to you later.

Reply:
The problem with warning is that with hundreds of agents being added, you will see hundreds of warning. With an exception, you only see 1.

Reply:
One exception and the model stops : )

I'm hesitant to use exception here because it isn't something really hazardous, like division by zero or missing attribute such that the application couldn't move forward.

Conversely it could be reasonable to do such conversion, as I mentioned above, when you have multiple shapefiles with different CRS. The user defines a common CRS in GeoSpace, then he/she is blocked by an error.

Would there be any way to *friendly* remind our users that such crs conversion is happening, and that they better take a look at it?

Reply:
> One exception and the model stops : )

This is recoverable by specifying the `auto_convert_crs`, and the user can then rerun the simulation. If the model is being run several times with warning, you will see hundreds of warning each time, and in the end you will have to specify a flag to hide the warning anyway. Though you can implement it in such a way that there is only 1 warning message, but this is assuming that all the agents in the input have identical crs.

Reply:
Yeah that's true. It is in fact more reasonable to have identical CRS for each GeoAgent type, since this is the case for GeoDataFrame, from where we create our GeoAgents. But until I figure out how to implement it, we're stuck with one CRS per GeoAgent object.

In terms of exception, would it make sense to be a `RuntimeError`?

Reply:
`ValueError` makes more sense to me.